### ISSUE TITLE
Performance - Reduce build time and memory usage

### ISSUE THREAD

**Issue by slorber** [facebook/docusaurus#4765]
## 💥 Proposal

With [Webpack 5 support](https://github.com/facebook/docusaurus/pull/4089), re-build times are now faster.

But we still need to improve the time for the first build which is not so good currently.

Some tools to explore:
- https://github.com/evanw/esbuild
- https://github.com/swc-project/swc
- https://github.com/alangpierce/sucrase

It will be hard to decouple Docusaurus totally from Webpack at this point. 

But we should at least provide a way for users to use an alternative (non-Babel) JS loader that could be faster and good enough. Docusaurus core should be able to provide a few alternate loaders that would work by default using the theme classic, by just switching a config flag.

If successful and faster, we could make one of those alternate loader the default loader for new sites (when no custom babel config is found in the project).

Existing PR by @SamChou19815 for esbuild: https://github.com/facebook/docusaurus/pull/4532

**Comment by slorber** [comment]
For anyone interested, we added the ability to customize the jsLoader here https://github.com/facebook/docusaurus/pull/4766

This gives the opportunity to replace babel by esbuild, and you can add this in your config:

```js
  webpack: {
    jsLoader: (isServer) => ({
      loader: require.resolve('esbuild-loader'),
      options: {
        loader: 'tsx',
        format: isServer ? 'cjs' : undefined,
        target: isServer ? 'node12' : 'es2017',
      },
    }),
  },
```

We don't document it yet (apart here) but may recommend it later for larger sites if it proves to be successful according to feedback from early adopters

**Comment by adventure-yunfei** [comment]
came from https://github.com/facebook/docusaurus/issues/4785#issuecomment-860705444.

Just wondering, is this issue aiming to reduce build time for entire site generator (including md/mdx parser, or [Docs](https://docusaurus.io/docs/docs-introduction)), or just [jsx react pages](https://docusaurus.io/docs/creating-pages)?

**Comment by slorber** [comment]
@adventure-yunfei md docs are compiled to React components with MDX, and the alternative js loader like esbuild also process the output of the MDX compiler, so this applies to documentation as well

**Comment by adventure-yunfei** [comment]
@slorber perfect! we're also trying to use esbuild to boost build time (for application project). I'll have a try on this.

**Comment by alphaleonis** [comment]
This gave a nice performance boost, although I think there are still more to be desired. Out of curiosity, what is actually going on that is taking so much time behind the scenes? In our case (a site with around 1,500 .md files) most of the time spent seems to be *before* and *after* the "Compiling Client/Compiling Server" progress bars appear and complete.

I just tested to add four versions to our site, and building it.  Before using esbuild, the process took just shy of **13 hours**(!). Using esbuild it was down to just shy of 8 hours. (Still way too long to be acceptable). So while it was a big improvement, it still seems to be very slow.

In the second case, it reported:
```
[success] [webpackbar] Client: Compiled successfully in 1.33h
[success] [webpackbar] Server: Compiled successfully in 1.36h
```

What was going on for the remaining 5 hours?   Is this normal behavior, or did we configure something incredibly wrong?

**Comment by slorber** [comment]
@alphaleonis it's hard to say without further analysis, but the MDX compiler is transforming each md doc to a React component, that is later processed by babel (or esbuild). 

The MDX compiler might be a bottleneck, this is why I'd like to provide an alternate MD parser for cases where MDX is not really needed. 

Webpack might also be a bottleneck.

Using esbuild is not enough. Also when using esbuild as a webpack loader, we are not really leveraging the full speed benefits of esbuild. Unfortunately we can't replace replace Webpack by esbuild easily, it's part of our plugin lifecycle API and Webpack is more featured than esbuild (we use various things like file-loader, svgr loader...)

> What was going on for the remaining 5 hours? Is this normal behavior, or did we configure something incredibly wrong? 

We have enabled Webpack 5 has persistent caching, and rebuild times are much faster. You need to persist `node_modules/.cache` across build to leverage it.

> And why does it take much longer than four times the amount of time with four versions added?

It's hard to tell without measuring on your system. Your system may have not enough memory for Webpack to do its job efficiently, leading to more garbage collection or whatever.

**Comment by alphaleonis** [comment]
@slorber Thanks for the explanation. We did try the persistent caching, and it seems to help a lot with the time spent during the "Build server/client" phase (which I assume is Webpack).  The machine in question had 16GB memory, and the same was specified as `max_old_space_size`.

Is there any way we can do some further analysis, such as enabling some verbose logging to get some more details perhaps?  Or is this kind of the expected build time for sites of that size?  (If so I guess we will have to find another solution for versioning, such as building/deploying each version separately.

**Comment by johnnyreilly** [comment]
> Also when using esbuild as a webpack loader, we are not really leveraging the full speed benefits of esbuild

This is true - but there's still a speed benefit to take advantage of. It's also pretty plug and play to make use of. See my post here:

https://blog.logrocket.com/webpack-or-esbuild-why-not-both/

**Comment by slorber** [comment]
> Is there any way we can do some further analysis, such as enabling some verbose logging to get some more details perhaps?

This is a Webpack-based app, and the plugin system enables you to tweak the Webpack config to your needs (`configureWebpack` lifecycle) and add logs or whatever you want that can help troubleshoot the system. You can also modify your local docusaurus and add tracing code if you need.

I'm not an expert in Webpack performance debugging so I can't help you much on how to configure webpack and what to measure exactly, you'll have to figure out yourself for now.

> Or is this kind of the expected build time for sites of that size?

It's hard to have meaningful benchmarks. Number of docs is one factor but also the size of docs obviously matter so one site is not strictly comparable to another. 40min build time for 2000 mdx docs with babel seems expected when comparing with other sites. Obviously it's too much and we should aim to reduce that build time, but it's probably not an easy thing to do.

> (If so I guess we will have to find another solution for versioning, such as building/deploying each version separately.

For large sites, it's definitively the way to go, and is something I'd like to document/encourage more in the future.
It's only useful to keep multiple versions in master when you actively update them. 
Once a version becomes unmaintained, you should rather move it to a branch and create a standalone immutable deployment for it, so that your build time does not increase as time passes and your version number increase.

We have made it possible to include "after items" in the version dropdown, so that you can include external links to older versions, and we use it on the Docusaurus site itself:

![image](https://user-images.githubusercontent.com/749374/123251352-631a4080-d4eb-11eb-8316-eb7782c9ba96.png)

I also want to have a "docusaurus archive" command to support this workflow better, giving the ability to publish a standalone version of an existing site and then remove that version.

**Comment by adventure-yunfei** [comment]
> Tested with a higher perf PC:
> 
> * with Docusaurus 2.0.0-beta.0, doc site generation finished in 63min
> * with [latest in-dev version](https://github.com/facebook/docusaurus/tree/master/website), doc site generation finished in 30min.
>   Reduced 50% time. 👍

Saddly the process costs a very large memory. 
My local testing environment has 32G memory, but in CICD environment memory limit is 20G. The process is killed cause of OOM, during emitting phase. From the monitor, the memory suddenly increased from 8G to 20G+.

**Comment by slorber** [comment]
It is unexpected that beta.2 is faster than beta.0, maybe you didn't clear your cache?

> The process is killed cause of OOM, during emitting phase. From the monitor, the memory suddenly increased from 8G to 20G+.

What do you mean by the "emitting phase"? I didn't take much time to investigate all this so any info can be useful.

**Comment by adventure-yunfei** [comment]
> It is unexpected that beta.2 is faster than beta.0, maybe you didn't clear your cache?

I'm using the esbuild-loader config from the docusaurus website example. So it should be esbuild making build faster.
 
> What do you mean by the "emitting phase"? I didn't take much time to investigate all this so any info can be useful.

This may not be accurate. The process memory was 7G most times. About 20 minutes later memory jumped to 20.2G while the console showing Client "emitting". After client build finished, the memory dropped down to 7G. (The Server was still building)

**Comment by krillboi** [comment]
Trying to test `esbuild-loader` but running into some trouble.

I have added the following to the top level of my `docusaurus.config.js` file:

```
  webpack: {
    jsLoader: (isServer) => ({
      loader: require.resolve('esbuild-loader'),
      options: {
        loader: 'tsx',
        format: isServer ? 'cjs' : undefined,
        target: isServer ? 'node12' : 'es2017',
      },
    }),
  },
```

I have added the following to my dependencies in `package.json`:

```
    "esbuild-loader": "2.13.1",
```

The install of `esbuild-loader` fails. Am I missing more dependencies for this to work? Might also be a Windows problem, unsure right now.

**Comment by krillboi** [comment]
Seems like it was one of the good ol' corporate proxy issues giving me the install troubles..

I'll try and test the `esbuild-loader` to see how much faster it is for me.

**Comment by krillboi** [comment]
Tested yesterday with production build, took about 3 hours compared to 6 hours before (`~400 docs x 5 versions x 4 languages`). 

So about half the time with the `esbuild-loader` which is nice. But we are reaching a size of docs where I am now looking into archiving older versions as seen on the Docusaurus site.

> This may not be accurate. The process memory was 7G most times. About 20 minutes later memory jumped to 20.2G while the console showing Client "emitting". After client build finished, the memory dropped down to 7G. (The Server was still building)

I witnessed the same thing where the memory usage would suddenly spike up to take 25+ gb.

**Comment by slorber** [comment]
Thanks for highlighting that, we'll try to figure out why it takes so much memory suddenly

**Comment by slorber** [comment]
Not 100% related but I expect this PR to improve perf (smaller output) and reduce build time for sites with very large sidebars: https://github.com/facebook/docusaurus/pull/5136 (can't really tell by how much though, it's site specific so please let me know if you see a significant improvement)

**Comment by adventure-yunfei** [comment]
> Not 100% related but I expect this PR to improve perf (smaller output) and reduce build time for sites with very large sidebars: #5136 (can't really tell by how much though, it's site specific so please let me know if you see a significant improvement)

Tested my application with latest dev version.
- Max memory usage: 21G
- Build time: 34min

Seems not working for my case.

**Comment by adventure-yunfei** [comment]
> This may not be accurate. The process memory was 7G most times. About 20 minutes later memory jumped to 20.2G while the console showing Client "emitting". After client build finished, the memory dropped down to 7G. (The Server was still building)

I've made another test, using plugin to override `.md` loader with noop:

```js
// inside docusaurus.config.js
{
  // ...
  plugins: [
    function myPlugin() {
      return {
        configureWebpack() {
          return {
            module: {
              rules: [
                {
                  test: /\.mdx?$/,
                  include: /.*/,
                  use: {
                    loader: require('path').resolve(__dirname, './scripts/my-md-loader.js')
                  }
                }
              ]
            }
          }
        }
      };
    }
  ],
}
```

```js
// scripts/my-md-loader.js
module.exports = function myPlugin() {
    const callback = this.async();
    return callback && callback(null, 'empty...');
};
```

And then run doc builder again.

- build time: 17min
- max memory: 20+G

So I'm afraid it's the code of page wrapper (e.g. top bar, side navigation, ...) that causes the max memory usage. Switching mdx-loader to another one may won't help.

**Comment by slorber** [comment]
@adventure-yunfei it's not clear to me how to do those measures, can you explain?

If you allow Docusaurus to take up to 20go, it may end up taking 20go. And it may take more if you give it more. The question is, how much can you reduce the `max_old_space_size` nodejs setting until it starts crashing due to OOM.

> So I'm afraid it's the code of page wrapper (e.g. top bar, side navigation, ...) that causes the max memory usage. Switching mdx-loader to another one may won't help.

Proving a memory issue is not the mdx-loader does not mean it's the "page wrapper". There is much more involved than the React server-side rendering here.

I suspect there are optimizations that can be done in this webpack plugin's fork we use: https://github.com/slorber/static-site-generator-webpack-plugin/blob/master/index.js

Gatsby used it initially and replaced it with some task queueing system.

**Comment by adventure-yunfei** [comment]
> Proving a memory issue is not the mdx-loader does not mean it's the "page wrapper". There is much more involved than the React server-side rendering here.

That's true. By saying "page wrapper" I mean any other code outside the md page content itself. Just trying to provide more perf information to help identify the problem.

More info:

- `max_old_space_size` was set as 4096.
- memory was monitored by windows perf monitor, with following result:
  ![image](https://user-images.githubusercontent.com/12392344/131658750-a8a60180-371d-42e2-948a-ba49c67bdaa5.png)
- the console stuck here for a long time:
  ![image](https://user-images.githubusercontent.com/12392344/131657932-53cdb5ef-a38b-4a1e-a11b-80c5703d8436.png)

**Comment by hjiog** [comment]
When the number of documents is large,run yarn start is still slow，do you have a plan to support vite?

**Comment by Josh-Cena** [comment]
The bundler is a bit hard to swap out. Next.js afaik is going through the same struggle, but basically the entire core infra is coupled with Webpack, so the most we can do is using different JS loaders (esbuild vs Babel) rather than letting you use an entirely different bundler. If you have the energy... you can try forking Docusaurus and re-implementing core with Vite.

**Comment by slorber** [comment]
There is some interest in making Docusaurus bundler & framework-agnostic in the future through an adapter layer but it's likely to be complex to implement in practice, and our current plugin ecosystem is also relying to Webpack so it would be a disruptive breaking change for the community.

**Comment by Josh-Cena** [comment]
Makes me wonder if it's possible to swap out Webpack in our core entirely🚎 As Docusaurus 3.0, rebuilt with Vite/Parcel/...

**Comment by armano2** [comment]
@slorber i did some big refactoring, small optimization and removed bunch of dependencies oo `static-site-generator-webpack-plugin` - https://github.com/slorber/static-site-generator-webpack-plugin/pull/2

i trimmed down package but there is still bunch of improvements to be done there

-------------------------------------

we should generally avoid using as this is "extremely" slow
```
const webpackStatsJson = webpackStats.toJson({ all: false, assets: true }, true);
```

**Comment by adventure-yunfei** [comment]
**I've found and fixed the large maximum memory issue in https://github.com/slorber/static-site-generator-webpack-plugin/pull/3.**

## Investigation

After investigation, the max memory happened during static-site-generator-webpack-plugin, rendering for every page path. So I took a look at static-site-generator-webpack-plugin code, and found two problems:

1. **memory/gc issue**. All pages are rendered during the same time (see [code](https://github.com/slorber/static-site-generator-webpack-plugin/blob/de03472bbb93aca73d2dc1dc7ecc812b8b5f3605/index.js#L64)). The render is async, and its allocated resource cannot be freed util the render promise finished. Thus **in the worst case, the maximum allocated memory should be sum of resources for rendering every page**, that is O(M*N) memory which M for page count and N for allocated memory size for rendering one page. That's not necessary.
2. **duplicate rendering**. After rendering one page, it'll crawl relative paths and continue render that relative path/page (see [code](https://github.com/slorber/static-site-generator-webpack-plugin/blob/de03472bbb93aca73d2dc1dc7ecc812b8b5f3605/index.js#L102)). That may lead to many duplicate page renderings.

To fix them:

- Using a promise queue, to render only partial pages in the same time, wait util finished, and then continue render next partial pages. Then we have maximum O(S*N) memory, which S for queue size, and N for allocated memory size for rendering one page.
- Record rendered pages, and skip rendering if duplicate.

## Check the optimization result below

#### Before optimization

- output logs (with some custom injections):
  ```ts
  [en] Creating an optimized production build...
  Fri Dec 03 2021 10:58:31 GMT+0800 (GMT+08:00): start compiling.
  i Compiling Client
  i Compiling Server
  Fri Dec 03 2021 11:07:15 GMT+0800 (GMT+08:00) start StaticSiteGeneratorWebpackPlugin. paths count: 7722
  Fri Dec 03 2021 11:07:19 GMT+0800 (GMT+08:00) after evaluate source. source size: 64794924
  Fri Dec 03 2021 11:21:24 GMT+0800 (GMT+08:00) renderPaths finished
  √ Client: Compiled successfully in 23.66m
  √ Server: Compiled successfully in 23.70m
  Fri Dec 03 2021 11:22:56 GMT+0800 (GMT+08:00): compile finished.
  Fri Dec 03 2021 11:22:56 GMT+0800 (GMT+08:00): start post build, plugins:
      docusaurus-plugin-content-docs, docusaurus-plugin-content-blog,
      docusaurus-plugin-content-pages, docusaurus-plugin-sitemap,
      docusaurus-theme-classic, docusaurus-bootstrap-plugin, docusaurus-mdx-fallback-plugin
  Fri Dec 03 2021 11:22:56 GMT+0800 (GMT+08:00): post build finished
  Fri Dec 03 2021 11:22:56 GMT+0800 (GMT+08:00): start handleBrokenLinks
  info Docusaurus found broken links!
  ...
  Fri Dec 03 2021 11:31:20 GMT+0800 (GMT+08:00): handleBrokenLinks finished
  Success! Generated static files in "build".
  ```
- memory records:
   ![before](https://user-images.githubusercontent.com/12392344/144550677-bcdd8d41-15fe-43e3-a88e-38012b998442.PNG)

The maximum allocated memory is **21+G**, increased quickly during static-site-generator-webpack-plugin `renderPaths`, and then dropped down quickly (from 11:07 to 11:21).

#### After optimization

- output logs:
  ```ts
  [en] Creating an optimized production build...
  Fri Dec 03 2021 16:15:06 GMT+0800 (GMT+08:00): start compiling.
  i Compiling Client
  i Compiling Server
  Fri Dec 03 2021 16:25:00 GMT+0800 (GMT+08:00) start StaticSiteGeneratorWebpackPlugin. paths count: 7722
  Fri Dec 03 2021 16:25:04 GMT+0800 (GMT+08:00) after evaluate source. source size: 64794924
  Fri Dec 03 2021 16:40:11 GMT+0800 (GMT+08:00) renderPaths finished
  √ Client: Compiled successfully in 25.70m
  √ Server: Compiled successfully in 25.83m
  Fri Dec 03 2021 16:42:00 GMT+0800 (GMT+08:00): compile finished.
  Fri Dec 03 2021 16:42:00 GMT+0800 (GMT+08:00): start post build, plugins: docusaurus-plugin-content-docs, docusaurus-plugin-content-blog, docusaurus-plugin-content-pages, docusaurus-plugin-sitemap, docusaurus-theme-classic, docusaurus-bootstrap-plugin, docusaurus-mdx-fallback-plugin
  Fri Dec 03 2021 16:42:00 GMT+0800 (GMT+08:00): post build finished
  Fri Dec 03 2021 16:42:00 GMT+0800 (GMT+08:00): start handleBrokenLinks
  info Docusaurus found broken links!
  ...
  Fri Dec 03 2021 16:50:52 GMT+0800 (GMT+08:00): handleBrokenLinks finished
  Success! Generated static files in "build".
  ```
- memory records:
  ![after](https://user-images.githubusercontent.com/12392344/144573981-35411bfe-d994-44ea-bbdd-0e3be4c874d5.PNG)

The maximum allocated memory is **7.1G** during static-site-generator-webpack-plugin `renderPaths` (from 16:25 to 16:40), without large memory allocated. (and maximum **8.2G** for the whole build, happens during docusaurus core `handleBrokenLinks`)).

The maximum memory decreased, while the build time remained the same.

## Further information

1. Build time summary, total 35min:
    - webpack compile: 27min
      - static-site-generator-webpack-plugin `renderPaths`: 15min
    - handleBrokenLinks: 8min
2. The render page function is defined in docusaurus core [`serverEntry.ts`](https://github.com/facebook/docusaurus/blob/95f911efefc96f78e1f2137ff65e7a2edda5cb62/packages/docusaurus/src/client/serverEntry.js#L42).  After checking the code:
    - I guess the large memory allocating comes from [minifier](https://github.com/facebook/docusaurus/blob/95f911efefc96f78e1f2137ff65e7a2edda5cb62/packages/docusaurus/src/client/serverEntry.js#L135). I've seen minifier consuming large memory before.
    - manifest is read & parsed multiple times (see [code](https://github.com/facebook/docusaurus/blob/95f911efefc96f78e1f2137ff65e7a2edda5cb62/packages/docusaurus/src/client/serverEntry.js#L109)). We can optimize it to only read once. 
3. In my case only large memory allocating issue is validated (@krillboi please help to validate in your case). There's no relative path in my case, so perf result for "avoid duplicate page rendering" is not validated.  @alphaleonis you may test it in your case. From your description I suspect your non-linear build time increasing is caused by this code.

**Comment by RDIL** [comment]
Tip: one of the best ways to reduce build time and memory is to use esbuild-loader instead of babel-loader.  See the website in this repo's config for the setup and use.

**Comment by slorber** [comment]
Thanks for working on this, will read all that more carefully and review PRs soon.

FYI afaik Gatsby also moved to a queuing system a while ago and that was something I wanted to explore. It's worth comparing our code to theirs.

---

Something I discovered recently: JS can communicate more seamlessly with Rust thanks to napi_rs with some shared memory, while it's more complicated in Go.
https://twitter.com/sebastienlorber/status/1460624240579915785
https://twitter.com/sebastienlorber/status/1468522862990536709

It's really worth trying to use SWC instead of esbuild with the Babel loader for that reason. 
I believe it may be faster than esbuild when used as a loader, while esbuild may be faster when you go all-in and stop using Webpack/loaders.
Next.js has great results with SWC.
https://nextjs.org/blog/next-12#faster-builds-and-fast-refresh-with-rust-compiler

**Comment by Josh-Cena** [comment]
@slorber I have a question that I'm unable to figure out: if our site is using esbuild, why is there still a Babel message in the command line saying that the changelog has exceeded 500KB?

**Comment by RDIL** [comment]
Personally, after using SWC and ESBuild for a while, I honestly prefer ESBuild.  SWC is not documented nearly as much, and ESBuild has very frequent releases fixing bugs and adding features.  ESBuild has a nicer DX IMO.

**Comment by alexander-akait** [comment]
swc and webpack are planned smoothly integration so if you want to avoid big changes/refactor/unpredictable bugs, prefer to use swc - https://github.com/swc-project/swc/tree/main/crates/swc_webpack_ast, also swc has better perf in some cases, but they both are fast.

swc is parser/codegen/visitor/bundler/transpiler/etc stuff, it is more just bundler, these are slightly different things, so if in the future you want deeper native integration, especial based on rust, I recommend to use swc

**Comment by RDIL** [comment]
ESBuild and swc's performance difference should be very tiny, given how, especially compared to JS tools, they are both much faster.  I don't really think its worth comparing the 2 for performance, since both are clearly very fast.

If one provides a better experience than the other, is it really worth benchmarking them on a base of like half a second?

**Comment by alexander-akait** [comment]
Please read:

>  also swc has better perf in some cases, but they both are fast.

I am more about - `swc` provides more things out of box, so if you need custom plugin/transformer/code generator for js/css/and more things, I strongly recommend to use `swc`, bundling is not only one things in build pipeline

**Comment by slorber** [comment]
Thanks @alexander-akait, we'll try to keep up with the work Vercel and Webpack are doing and see what we can reuse here.

SWC is more extensible and we may even implement some Rust plugin someday to process our i18n API and replace the global registry of messages with some inline localized translation strings directly in the app bundle. (ie, better code splitting for translations).

And it should also make it easier to use Emotion/StyledComponents with Docusaurus, as Vercel is already working on porting existing Babel plugins to Rust.

> @slorber I have a question that I'm unable to figure out: if our site is using esbuild, why is there still a Babel message in the command line saying that the changelog has exceeded 500KB?

😅 good question, maybe it's related to the translation extraction? But afaik it's not run when starting the dev server... weird

**Comment by Josh-Cena** [comment]
A quick guess is that MDX v1 uses Babel under the hood to do the transformation: https://github.com/mdx-js/mdx/blob/master/packages/mdx/mdx-hast-to-jsx.js#L1

It seems MDX v2 has removed this dependency

**Comment by yangshun** [comment]
I've been using Next.js a fair bit this year and honestly if I could turn back time, I think @endiliey and I shouldn't have built our own site generator in Docusaurus v2 and we should have used Next.js instead. Admittedly, it was a mix of not-invented-here syndrome and wanting to learn how to build a site generator from scratch.

At this point, Next.js is a clear winner in the SSG race and Gatsby is more or less out. Vercel is doing so well with their latest funding rounds and star hires, I think it's safe to bet on Next.js.

Docusaurus v2 is split into 3 layers: our homegrown (1) SSG infra, (2) plugins, (3) UI/themes. If I were to build Docusaurus v3, I would make it such that Docusaurus 3 is more like [Nextra](https://github.com/shuding/nextra/), swap out (1) with Next.js and retain (2) and (3). Docusaurus 3 would provide all the documentation-related features. I felt that Docusaurus 2 had to play catch up a lot and implement lots of non-documentation-specific features that were required by websites when Next.js already did all these. We could have saved lots of time by riding on the shoulders of giants.

With Next.js' current popularity and trajectory, I think it's only a matter of time before someone builds a fully-fledged docs theme on top of Next.js that does everything that Docusaurus does, but probably better because their SSG infra is much more optimized by virtue of being on Next.js. IMO many users would also like to have the SSR features Next.js provides so that they can build auth have better integration with their core product.

**Comment by Josh-Cena** [comment]
I still like the idea of having "dependency independence". Apart from Webpack / React router / other low-level infra, we aren't coupled to any dependency. It means we can describe our architecture as an integral thing without saying "the peripheral is Docusaurus, but the core, well, is Next.js and it's a black box". Working on Docusaurus frankly made me a lot more familiar with how SSG works😄

**Comment by slorber** [comment]
@yangshun @Josh-Cena we seem to all agree on this

That was also my opinion on day one, but also think that having our own SSG wasn't totally useless: it permitted us to iterate faster without being blocked by limits of an existing dependency and gave us time to evaluate better Gatsby vs Next.js vs others.

We discussed this with @zpao and @JoelMarcey a few months ago and we agreed that Docusaurus should rather migrate to Next.js.

Or become framework-agnostic. This might be more complicated to document well, and harder to implement, but could allow using other solutions like Remix or Gatsby.

And building on top of Next.js also incentives Vercel to invest in Docusaurus 🤷‍♂️ eventually we could join forces with Nextra

**Comment by Josh-Cena** [comment]
One thing I regret about migrating to Next.js is we will be forever tied to Webpack because from my observation the Webpack 5 migration for them was more painful than for us. Webpack ultimately is not comparable in terms of performance to, say, esbuild... 🤔

**Comment by alexander-akait** [comment]
Next.js start to migrate on swc (rust), and replace webpack more and more, so you should not afraid it

**Comment by slorber** [comment]
> One thing I regret about migrating to Next.js is we will be forever tied to Webpack because from my observation the Webpack 5 migration for them was more painful than for us. Webpack ultimately is not comparable in terms of performance to, say, esbuild... 🤔

I agree that we want something fast but I believe it's also the goal of Next.js 😅 

Their Webpack 5 migration is likely more complex because of the higher diversity of sites needing to migrate, compared to our low diversity: most doc sites are not customized that much and plugins don't always tweak Webpack settings.

Also, there's value in keeping at least some things in Webpack for now: our plugin ecosystem can remain retro-compatible

**Comment by Josh-Cena** [comment]
Yeah, in the short term migrating to Next.js is surely going to yield lots of benefits. I'm never actually used it purely as an SSG but more as a React framework, but if we can figure out how to make them interoperate it will be very nice!

**Comment by yangshun** [comment]
> Webpack ultimately is not comparable in terms of performance to, say, esbuild

The thing is, with Next.js' backing, they will just use the fastest that's out there and we can benefit from it by building on top of Next.js. I believe Sebastien is also saying the same. Hopefully we can go with Next.js in the next version (or even better if can be framework agnostic)!

**Comment by gabrielcsapo** [comment]
As an outsider we looked at all the options and docusaurus was the best when it came to level of investment and clean and clear plugin and theming architecture. I think competition in this space is much needed and I think having nextra and docusaurus is great for pushing the envelope. 

I think the alternatives to webpack aren't somewhere stable enough to really compare apples to apples, I think by the next major version I think the landscape is going to look very different or maybe webpack migrates to rust and no one needs to do any major rearchitecting at all.

**Comment by Josh-Cena** [comment]
Allowing alternative JS loaders may hinder the provision of useful OOTB JS syntax extensions. For example, @slorber mentioned somewhere that we may explore making `<Translate>` a zero-runtime API through Babel transformations. I also talked to him about solving #4530 through a runtime transformation of `'@theme-original/*'` to the actual path of "next component in the theme component stack", instead of using static Webpack aliases. I would definitely want to use SWC/esbuild, but in any case, it would mean writing the transform plugin with a different set of APIs, maybe even a different language. That makes it not scalable. If we have to insert an extra loader that uses Babel, then we are back in square one and perf will be compromised.

**Comment by RDIL** [comment]
Perf is arguably already compromised by using Babel in the first place.

**Comment by RDIL** [comment]
I think we should drop Babel personally.

Pros:
- Better perf + memory usage
- Smaller dependency tree, causing potentially faster install times

Cons:
- Transforms may need to be ported to Rust, not 100% sure on that one though.
- Would need to upgrade to MDX 2?

**Comment by slorber** [comment]
> Transforms may need to be ported to Rust, not 100% sure on that one though.

If we want i18n to work we'd rather port our plugin to Rust yes.

Although technically for now we only extract translations with the Babel plugin (no source change), so I think it may not even be necessary in the short term: we could keep extracting translations with Babel (slower but who cares) and only transpile in Rust.

> If you are talking about SWC—the current plugin system is still JS-based. [swc.rs/docs/usage/plugins](https://swc.rs/docs/usage/plugins)

Afaik NAPI-RS has low overhead to call RS from JS but not the opposite. That's also probably why Vercel (recently hired NAPI-RS creator) is porting popular Babel plugins to Rust

**Comment by Josh-Cena** [comment]
> Although technically for now we only extract translations with the Babel plugin (no source change), so I think it may not even be necessary in the short term: we could keep extracting translations with Babel (slower but who cares) and only transpile in Rust.

Yeah, that's the idea. The extractor is only run in development so it's fine to be in Babel, but for build-time transformations (if we ever implement that) we'd rather use Rust.

**Comment by fawazahmed0** [comment]
Hi, I am trying to use github actions to build my large site, currently github actions provide 6 hours and after that it timeouts.
Is there any option to save the partial build of docusaurus site and resume the build again.

**Comment by slorber** [comment]
We don't have a way to resume a build sorry.

We use Webpack which has caching layers (that we can persist) but afaik it can only persist at the end of a build and not incrementally. Unless someone shows how to persist Webpack cache before the end of a build, I assume it's not possible

**Comment by Josh-Cena** [comment]
I've thought about something similar in the past, using some sort of "lazy build" that doesn't load everything at once into memory. I don't know if Webpack is able to do that.

**Comment by alexander-akait** [comment]
Maybe useful https://webpack.js.org/configuration/experiments/#experimentslazycompilation

**Comment by Josh-Cena** [comment]
Ah yes! I did see that. Definitely worth looking into in the near future...

**Comment by slorber** [comment]
@alexander-akait already tried that but wasn't able to make it work successfully so far 🤪 

<img width="404" alt="CleanShot 2022-05-25 at 19 55 26@2x" src="https://user-images.githubusercontent.com/749374/170331064-c6776ec4-f72f-4791-a218-6d6ed8b17d54.png">


Also is this really supposed to improve a static site production build? Considering that in the end, everything must be compiled what's the point of using lazy compilation?

This Storybook benchmark with/without lazyCompilation also shows that the win for a production build is not significant: https://storybook.js.org/blog/storybook-performance-from-webpack-to-vite/

<img width="672" alt="CleanShot 2022-05-25 at 19 54 00@2x" src="https://user-images.githubusercontent.com/749374/170330769-874a7aac-eeb9-4596-aa70-f9753fc130af.png">

But it is useful for the dev env:

<img width="717" alt="CleanShot 2022-05-25 at 19 55 13@2x" src="https://user-images.githubusercontent.com/749374/170330987-76d0e606-7d8b-4f0d-8340-2788d08551ee.png">

---

We definitively want the improvements of lazyCompilation, but this issue is more about the production build, so I'm not sure this option is very relevant here?

**Comment by alexander-akait** [comment]
@slorber I see, currenty my reccomendation is trying to switch from babel to swc (and swc minifier) too, it should help very well, also swc is more mature now

**Comment by fawazahmed0** [comment]
If you get a `terminate called after throwing an instance of 'std::bad_alloc'` error during the build, you might also need to set `vm.max_map_count` to higher value. for example `sysctl -w vm.max_map_count=655300` 

[Reference](https://github.com/nodejs/node/issues/27715#issuecomment-578557226)

**Comment by slorber** [comment]
FYI the current canary and next release will include the suggestion from @adventure-yunfei 

We now limit the concurrency when outputting static files at the end of the build process. This should overload less the system IOs and reduce memory footprint by default.

Note: this may not impact build time much, but mostly avoid a potential OOM at the end of your build.

---

There's a "secret" env variable in case you want to tune this concurrency setting: `process.env.DOCUSAURUS_SSR_CONCURRENCY`

https://github.com/facebook/docusaurus/pull/7547/files#diff-058c5cef3799e9df200345c9d5b769bbc27838aea32aa4f8345d09858d416781R96


For now, this is undocumented: please give us feedback on how impactful it is on your site, and eventually, we'll document it or add it as a first-class config setting.

**Comment by vladfrangu** [comment]
Just tried that secret env variable (on version `0.0.0-5101`) and even setting it to `2` I get an OOM error on our website (I can link the repository if needed). Is there anything I can do to debug why the memory usage spikes up so high once the server is supposedly compiled?

**Comment by slorber** [comment]
> Just tried that secret env variable (on version 0.0.0-5101) and even setting it to 2 I get an OOM error on our website (I can link the repository if needed). Is there anything I can do to debug why the memory usage spikes up so high once the server is supposedly compiled?

@vladfrangu consuming memory, above the default available in Node.js (0.5gb), does not seem unexpected to me for a large site. Now if we can't reasonably build a large site with 2-10GB of memory, that seems way more problematic.

The only thing you can do is profile your build and report your findings, to know which step exactly is taking more memory than you expect. I can't really teach you how to do this through a GitHub issue, I'm not an expert in this either.

See @fawazahmed0 comment here: https://github.com/facebook/docusaurus/issues/4765#issuecomment-910164698

Having a curve + an idea of what Webpack is working on is helpful.

**Comment by Ornias1993** [comment]
@alexander-akait 
Do you have exaples of using SWC and SWC minifier instead of babel?

**Comment by slorber** [comment]
> @alexander-akait Do you have exaples of using SWC and SWC minifier instead of babel?

we are using SWC on the Docusaurus site now, so feel free to steal our config: https://github.com/facebook/docusaurus/pull/6944

**Comment by wenerme** [comment]
Possible to use mdx esbuild loader https://mdxjs.com/packages/esbuild/  to speedup the md build ?

**Comment by rxliuli** [comment]
刚从 vite-vuepress 迁移过来，但我认为 docusaurus 还有很大的性能提升空间，一个基本构建性能对比。

docusaurus

```sh
$ time pnpm docs-build

> to-the-stars@1.0.0 docs-build C:\Users\rxliuli\Code\book\to-the-stars
> docusaurus build

[INFO] [zh-Hans] Creating an optimized production build...

✔ Client
  

✔ Server
  Compiled successfully in 39.38s


✔ Client


● Server █████████████████████████ cache (99%) shutdown IdleFileCachePlugin
 stored

[SUCCESS] Generated static files in "build".
[INFO] Use `npm run serve` command to test your build locally.

real    0m51.324s
user    0m0.030s
sys     0m0.075s
```

vite-vuepress

```sh
$ time pnpm docs-build

> to-the-stars@1.0.0 docs-build C:\Users\rxliuli\Code\book\test-tts
> pnpm docs-setup && vuepress build && rimraf .vuepress/dist/books/


> to-the-stars@1.0.0 docs-setup C:\Users\rxliuli\Code\book\test-tts
> tsx .vuepress/init.ts

⠼ Initializing and preparing dataindex start 82
index end 82
⠴ Initializing and preparing dataindex write
✔ Initializing and preparing data - done in 2.55s
✔ Compiling with vite - done in 10.69s
✔ Rendering 85 pages - done in 2.19s
✔ Generating sitemap to sitemap.xml - done in 7ms
success VuePress build completed in 16.11s!

real    0m18.996s
user    0m0.030s
sys     0m0.045s
```

文档站点的大概信息

```ts
$ cloc books/
      83 text files.
      83 unique files.
      15 files ignored.

github.com/AlDanial/cloc v 1.94  T=0.52 s (158.4 files/s, 112091.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Markdown                        83          29090              0          29645
-------------------------------------------------------------------------------
SUM:                            83          29090              0          29645
-------------------------------------------------------------------------------
```

不确定具体原因，但猜测可能包括下面几个

1. 使用 babel 而不是 esbuild/swc 这种原生编译器
2. 使用 webpack 而非更好的 bundle，例如 rollup？
3. 使用 mdx 而非 markdown-it（我很喜欢 mdast 系列的工具包，但它的 npm 包太过零散，很难相信它的性能是比较好的）

**Comment by Josh-Cena** [comment]
@rxliuli MDX is what we base our product on and advertise for. It is not replaceable. The JS loader is already customizable and our website uses SWC ourselves, and esbuild also works. Webpack _could_ be replaced, but it's too much work, and we have a plugin lifecycle literally called `configureWebpack` that would break if the config it injects becomes no longer compatible.

**Comment by wenerme** [comment]
Wow, 51s is so fast, My ci https://github.com/wenerme/wener/actions runs about 10-15m using esbuild loader, the loader only affects js/ts, but all docs are md, md parsing seems slow.

**Comment by rxliuli** [comment]
Using esbuild-loader, although it is still about twice as slow as vite-vuepress, it has improved some

```sh
$ time pnpm docs-build

> to-the-stars@1.0.0 docs-build C:\Users\rxliuli\Code\book\to-the-stars
> docusaurus build

[INFO] [zh-Hans] Creating an optimized production build...

✔ Client
  

✔ Server
  Compiled successfully in 27.10s


✔ Client
  

● Server █████████████████████████ cache (99%) shutdown IdleFileCachePlugin
 stored

[SUCCESS] Generated static files in "build".
[INFO] Use `npm run serve` command to test your build locally.

real    0m37.529s
user    0m0.031s
sys     0m0.060s
```

**Comment by alexander-akait** [comment]
For CSS and HTML minification you can use https://www.npmjs.com/package/@swc/css (https://github.com/webpack-contrib/css-minimizer-webpack-plugin) and https://www.npmjs.com/package/@swc/html (docs https://github.com/webpack-contrib/html-minimizer-webpack-plugin#swchtml)

It is very fast, but young (but all bugs are fixing very fast), so feel free to check it.

In near future we will provide ability to replace `css-loader` and `postcss-loader` (except non standard postcss-plugin, we will support all popular plugins) on `swc-loader`

**Comment by JakeChampion** [comment]
> There's a "secret" env variable in case you want to tune this concurrency setting: `process.env.DOCUSAURUS_SSR_CONCURRENCY`
> 
> [#7547 (files)](https://github.com/facebook/docusaurus/pull/7547/files#diff-058c5cef3799e9df200345c9d5b769bbc27838aea32aa4f8345d09858d416781R96)
> 
> For now, this is undocumented: please give us feedback on how impactful it is on your site, and eventually, we'll document it or add it as a first-class config setting.

Howdy ✌️ - I thought I'd give some feedback - I have a website with xx,xxx pages and was getting out-of-memory issues when building. Using a combination of swc-loader and `DOCUSAURUS_SSR_CONCURRENCY=5` has solved the problem and now it successfully builds ☺️

**Comment by RDIL** [comment]
Yeah, Babel is the root cause here, the problem is that the maintainers don't want to remove it because that would be a breaking change.

**Comment by Ornias1993** [comment]
> For CSS and HTML minification you can use https://www.npmjs.com/package/@swc/css (https://github.com/webpack-contrib/css-minimizer-webpack-plugin) and https://www.npmjs.com/package/@swc/html (docs https://github.com/webpack-contrib/html-minimizer-webpack-plugin#swchtml)
> 
> It is very fast, but young (but all bugs are fixing very fast), so feel free to check it.
> 
> In near future we will provide ability to replace `css-loader` and `postcss-loader` (except non standard postcss-plugin, we will support all popular plugins) on `swc-loader`
> 
> In plans - replace MDX (and provide markdown parser/transformers/codegen too), currently no date
> 
> Also for `svg` we are preparing `swc_xml_parser` and `swc_svg_minifier` (replacement for SVGO)

Any experience with this and docusaurus?

**Comment by Ornias1993** [comment]
> > There's a "secret" env variable in case you want to tune this concurrency setting: `process.env.DOCUSAURUS_SSR_CONCURRENCY`
> > [#7547 (files)](https://github.com/facebook/docusaurus/pull/7547/files#diff-058c5cef3799e9df200345c9d5b769bbc27838aea32aa4f8345d09858d416781R96)
> > For now, this is undocumented: please give us feedback on how impactful it is on your site, and eventually, we'll document it or add it as a first-class config setting.
> 
> Howdy ✌️ - I thought I'd give some feedback - I have a website with 9,785 pages and was getting out-of-memory issues when building. Using a combination of swc-loader and `DOCUSAURUS_SSR_CONCURRENCY=5` has solved the problem and now it successfully builds ☺️

swc loader also halved(!) the compilation time for our thousands of pages :)

**Comment by johnnyreilly** [comment]
[I wrote up swapping out babel-loader for swc-loader](https://blog.johnnyreilly.com/2022/09/29/faster-docusaurus-build-swc-loader). I might add in the `DOCUSAURUS_SSR_CONCURRENCY` tip too - might get us some more feedback!

**Comment by jhaals** [comment]
Just wanted to echo what's said here before about slow build speeds and taking up too much memory. We are using docusaurus for https://backstage.io but are struggling a lot with slow build speeds of almost an hour. We have switched to  swc-loader and using DOCUSAURUS_SSR_CONCURRENCY 5 as suggested by @johnnyreilly above but we are still have issues with OOM issues when allocating 7GB ram for the build process. We would love to use the versioned docs feature but I don't see how that's possible at this point.

Our settings can be found here for anyone interested https://github.com/backstage/backstage/blob/master/microsite-next/docusaurus.config.js#L66
https://github.com/backstage/backstage/blob/master/.github/workflows/deploy_microsite.yml

**Comment by Ornias1993** [comment]
@jhaals Seem with our TrueCharts documentation system.
We also have to cut content/pages because of the completely unusable build-times even with significant optimalisations such as jwc-loader and the SSR_CONCURRENCY setting

**Comment by rxliuli** [comment]
@jhaals @Ornias1993  Just a sharing, regarding the static document generator, there is an order of magnitude gap between vitepress and vuepress/docusaurus (in the actual test, the performance gap is 20 times), hoping that it can prompt docusaurus to do real  Performance improvements.
 a test effect

vitepress	1m56.019s
vuepress	14m18.764s
docusaurus	36m39.857s

```sh
$ cloc docs/
     914 text files.
     914 unique files.
       0 files ignored.

github.com/AlDanial/cloc v 1.94  T=2.60 s (351.5 files/s, 319491.6 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Markdown                       914         371604              0         459249
-------------------------------------------------------------------------------
SUM:                           914         371604              0         459249
-------------------------------------------------------------------------------
```

**Comment by Ornias1993** [comment]
100% agreed, this needs a major release with performance refactor soon(tm).
I also doubt this is really the performance facebook(inc) wants to showcase to the world.

It's in no-ones interest to have docusaurus perform this badly on bigger sites.

**Comment by slorber** [comment]
> I also doubt this is really the performance facebook(inc) wants to showcase to the world.
> It's in no-ones interest to have docusaurus perform this badly on bigger sites.

I feel like you don't understand how open-source works here. Facebook tries first to solve its own internal use-case and is generous to expose their tool to the world, for free. They don't owe you anything. You can fix the performance problems yourself by forking the project, or give us the bandwidth to do so by contributing financially to OpenCollective. Many big companies are using Docusaurus today, but none of them outside Facebook is contributing significantly to its development apart Meta. 

Please try to be part of the solution instead of complaining, and understand that Meta doesn't have an infinite budget to spend on Docusaurus alone just to please you.

**Comment by slorber** [comment]
Back to the performance issues. 

Unfortunately, I understand the frustration, but I have limited bandwidth to work on it in the short term because I have to handle the regular maintenance of the project, ensure pending PRs do not become stale, and work on the MDX 2.0 + React 18 upgrade for Docusaurus v3 before doing starting any other perf/infra work.

The MDX 2.0 upgrade should bring some perf gains because the tool is faster (see their release blog post). Also we'll remove the need to "retranspile" the mdx docs with Babel, this is an extra step we did historically but this definitively can be removed.

Regarding the memory consumption, out-of-memory errors, and things like that, I don't really think it's related to MDX but is probably something else in our system (the SSG part?) that we'll have to profile and optimize. If you are a good perf profiler, you can help investigate here and help us figure out what are the bottlenecks of Docusaurus currently.

I am 100% aware of the new wave of Rust tooling being released, I track them all and their updates in my newsletter [This Week In React](https://thisweekinreact.com). We'd like to explore the usage of various tools and adopt Rust at multiple levels. 

Good candidates to modernize our infrastructure:
- [SWC](https://swc.rs/) to transpile code (but all Docusaurus users with Babel plugins will be able to use it easily?)
- [mdxjs-rs](https://github.com/wooorm/mdxjs-rs) to compile MDX docs (but it doesn't have remark plugin support)
- [LightningCSS](https://github.com/parcel-bundler/lightningcss) to handle... the CSS
- [Turbopack](https://turbo.build/pack) to replace Webpack (recently announced loader/alias similar to Webpack, easier to adopt)
- [rspack](https://www.npmjs.com/package/@rspack/core): new WIP Rust port of Webpack (compatible) being worked on by ByteDance devs
- [Unplugin](https://github.com/unjs/unplugin): a modular system so that Docusaurus plugin authors can add bundling features and keep the plugins portable/bundler-agnostic at the same time
 
As you notice, many of these projects are quite new, not battle-tested yet, or will disrupt our plugin ecosystem in their current form. I think we can try to adopt them progressively, but it's also important to let them mature a bit. We will try to adopt these tools incrementally over time.

There's also the possibility to migrate Docusaurus to another infrastructure (like Next.js, Astro, Remix) and/or the attempt to make the opinionated features we provide framework-agnostic.

I hope you understand better the direction I want to take here. We'll figure out how to make Docusaurus fast in the long term, but it's unlikely it will become extremely fast in the very short term unfortunately.

**Comment by RDIL** [comment]
@slorber Babel and Terser are the biggest bottlenecks by far.  I've gotten it to build a 100000 page site in less than 10 minutes with 3gb of ram with terser and babel both switched out for esbuild

**Comment by Ornias1993** [comment]
> I feel like you don't understand how open-source works here. Facebook tries first to solve its own internal use-case and is generous to expose their tool to the world, for free. 

Please don't assume what I do ordo not know.

> They don't owe you anything. You can fix the performance problems yourself by forking the project, or give us the bandwidth to do so by contributing financially to OpenCollective. Many big companies are using Docusaurus today, but none of them outside Facebook is contributing significantly to its development apart Meta.

In stark contrast to Meta, I do not have near-unlimited money to throw at a virtual-wall and I'm already maintaining one opensource projects nearly full-time. Nor is the backend of docusaurus within my general development area of expertise, so the amount of hours required for me would not be beneficial

> Please try to be part of the solution instead of complaining, and understand that Meta doesn't have an infinite budget to spend on Docusaurus alone just to please you.

I never said I wanted to be pleased, I gave my opinion in how this looks like, to me as an outsider.
I see a failing company, have a docs product that underperforms, that is presented like a good performer on the website.

What I wanted to make clear, which I literally said, is that this is in the interest of no one. Not us as users, not you as maintainers nor meta as a company.

Not every comment on an issue is inherently aimed at solving the issue.
Is my opinion above constructive, no definitely not.
But it does showcase what this years old stagnated issue looks like from an outside perspective.

**Comment by Ornias1993** [comment]
> @slorber Babel and Terser are the biggest bottlenecks by far. I've gotten it to build a 100000 page site in less than 10 minutes with 3gb of ram with terser and babel both switched out for esbuild

Was that with docusaurus, or a similair DIY setup?
Some example code you care to share? :)

**Comment by RDIL** [comment]
With Docusaurus - https://github.com/facebook/docusaurus/discussions/3132#discussioncomment-3615781 for esbuild, can't remember exactly what it's called but there is an environment variable to disable Terser (`NO_MINIFY=true` I think?)

**Comment by Ornias1993** [comment]
> Regarding the memory consumption, out-of-memory errors, and things like that, I don't really think it's related to MDX but is probably something else in our system (the SSG part?) that we'll have to profile and optimize. If you are a good perf profiler, you can help investigate here and help us figure out what are the bottlenecks of Docusaurus currently.

Imho this is the biggest issue, besides generation times.
What I noticed personally, is that it's primarily bigger pages that cause the most ram consumption.

The different between 3000 small and 3000 big md pages, is insanely big.
We're talking about "10gb can do it" or "will eat up 40gb with ease and OOM halfway through". But still, within 10gb I can WITH EASE have 10 copies of the data in ram already.

It almost looks like a memoryleak to be honest.

It's also interesting to see errors about an internal babel dependency throwing "bigger than 500kb" warnings on bigger md files that are definately still under 200kb. So somewhere in the stack it looks like pages are blown up and not cleaned-up when processed correctly.

**Comment by RDIL** [comment]
> It's also interesting to see errors about an internal babel dependency throwing "bigger than 500kb" warnings on bigger md files that are definately still under 200kb. So somewhere in the stack it looks like pages are blown up and not cleaned-up when processed correctly.

That's MDX.

**Comment by Ornias1993** [comment]
> > It's also interesting to see errors about an internal babel dependency throwing "bigger than 500kb" warnings on bigger md files that are definately still under 200kb. So somewhere in the stack it looks like pages are blown up and not cleaned-up when processed correctly.
> 
> That's MDX.

Ahh yes, the above transpile refered.
Any sign if that's also the memoryleak (I think we can safely call it that at this point) is comming from?

**Comment by RDIL** [comment]
It's functionally leaking because the bundling process is going through thousands of files and trying to understand what code uses what modules, and there are hundreds of modules to manage, etc etc etc

**Comment by armano2** [comment]
I have been investigating this a little, generally webpack build is defenitly big part, that includes all plugins, but i was surprised to see that it actually took significant amount of time to generate webpack config. 

on my machine this roughly takes 1min from start of `buildLocale` method to execution of compile on webpack. 

<details>
  <summary>my spec</summary>
  
CPU: 12th Gen Intel(R) Core(TM) i7-12700H
Memory: 32.0 GB, 4800 MHz DDR5
SSD: Samsung SSD 980 PRO 1TB
</details>

as a test case i have been using this project

config generated by webpack seem substantial, as it contains 6k lines (https://gist.github.com/armano2/a7cf275c8763ab33c112ec0dc0269295)

----------

this may seem insignificant, but such big webpack config + time that it takes to generate it most likely is getting even slower/bigger for larger projects

**Comment by Ornias1993** [comment]
> It's functionally leaking because the bundling process is going through thousands of files and trying to understand what code uses what modules, and there are hundreds of modules to manage, etc etc etc
> 
> Each of those JS files need to be parsed into ASTs, optimized, tree-shaken, it's just an operation that is inefficient to do in JS itself, which is why Babel is not that great.

Understandable, but even so when GC cannot shrink ram consumption within the limits, it should differ ton”!(preferably high iops) disk.

tests with swap indicates that works relatively well from a performance perspective

**Comment by armano2** [comment]
> Understandable, but even so when GC cannot shrink ram consumption within the limits, it should differ ton”!(preferably high iops) disk.
> 
> tests with swap indicates that works relatively well from a performance perspective

actually there is more issues, build process requires to find all files and read them twice,

1. loop trough files and collect metadata, this is used to register all routes and feed to likify
2. webpack has to loop trough all files and build them with metadata provided

as metadata i'm talking about `---` and file names

---------

in theory we could move this to webpack plugin and collect them as they come while we are processing files but this will ahve to be `tapped` before processing of files

**Comment by Ornias1993** [comment]
> > Understandable, but even so when GC cannot shrink ram consumption within the limits, it should differ ton”!(preferably high iops) disk.
> > tests with swap indicates that works relatively well from a performance perspective
> 
> actually there is more issues, build process requires to find all files and read them twice,
> 
> 1. loop trough files and collect metadata, this is used to register all routes and feed to likify
> 2. webpack has to loop trough all files and build them with metadata provided
> 
> as metadata i'm talking about `---` and file names
> 
> in theory we could move this to webpack plugin and collect them before we process files,
> 
> webpack has a lot of `hooks` that we can `tap` into in plugins
> 
> with that we should reduce amount of iops by substantial number

Thats not what I was saying, I was saying that it should overflow to disk if there isnt enough ram.

we can already fake this with swap which works fine.

**Comment by sibelius** [comment]
what about cache?

can we cache the latest build output and only rebuild what changed?

**Comment by slorber** [comment]
@sibelius you can cache `node_modules/.cache` across builds, which should speed up the Docusaurus rebuilds thanks to Webpack 5 caching

Caching `build` and `.docusaurus` won't lead to any significant benefit, but I'll look into later to see if there are low-hanging fruit optimizations we could apply.

**Comment by mattrunyon** [comment]
Switching to swc-loader greatly helped our build times. We tried caching in Github Actions, but it was actually problematic for us because the cache size was 3.5GB and quickly got evicted. We probably could have set up our actions better to only create caches from the main branch builds, and only read caches for deploy preview builds on PRs.

It would be nice to be able to specify compression for the cache as well as a no-cache option. Webpack supports gzip and brotli for cache compression. While using brotli compression added some time to my local clean build (8:22 to 8:46 user time), it also cut the cache from 3.5GB to 1.5GB.

For people not using the cache in CI, disabling cache would be nice. This reduced my local build from 8:22 to 7:07 user time.

Minifying with esbuild is also a huge boost. 13:13 to 8:37. Only 15 seconds more than no minification at all for my build

Seems like there's some easy opt-in optimizations which should result in equivalent builds, but faster
- Option to minify with esbuild (or swc, but esbuild was faster for me). These have significant time savings. Would require the user to install esbuild as a dev dependency (assuming it's added to docusaurus as an optional peer dep)
- Option to disable cache. Can save a good chunk of time if not using cache
- Option to compress cache. Adds a bit of time, but can cut cache by 50+%

**Comment by slorber** [comment]
Thanks for reporting these low hanging fruits @mattrunyon 

FYI you should be able to disable cache or customize `cache.compress` by creating a plugin with `configureWebpack` lifecycle hook

**Comment by mattrunyon** [comment]
For anybody who wants a plugin to switch minifiers or disable/compress cache, here's an example

```js
{
    name: 'webpack-docusaurus-plugin',
    configureWebpack(config, isServer, utils) {
      // Disable cache in CI since it gets evicted too quickly from github actions limits
      const isCI = process.env.CI;
      const cacheOptions = isCI ? { cache: false } : {};

      // Or compress the cache
      // const cacheOptions = isCI ? { cache: { compression: 'brotli' } } : {};

      // Replace terser with esbuild minify, but only if terser would have been used
      // This still respects the --no-minify flag
      const minimizer = new TerserPlugin({
        minify: TerserPlugin.esbuildMinify,
      });
      const minimizers = config.optimization.minimizer?.map((m) =>
        m instanceof TerserPlugin ? minimizer : m
      );

      return {
        mergeStrategy: { 'optimization.minimizer': 'replace' },
        optimization: {
          minimizer: minimizers,
        },
        ...cacheOptions
      };
    },
  };
```

**Comment by SubJunk** [comment]
> please let us know if that works for your use-case.

Using `esbuild-loader` as described here https://github.com/facebook/docusaurus/issues/4765#issuecomment-841135926 results in my builds taking `12s` per language, down from about `14s` per language, so the difference is not big (maybe it was at the time of the post). For my small website https://support.universalmediaserver.com a production build on Cloudflare takes `19m 26s` total:
<img width="661" alt="image" src="https://github.com/facebook/docusaurus/assets/584228/9a137a49-b384-4f1a-8897-60df62ce45b1">

One thing I notice while watching the output logs, is that most of the time is spent re-compiling the dependencies over and over, for each language:
<img width="809" alt="image" src="https://github.com/facebook/docusaurus/assets/584228/0bd08771-d074-4d70-8776-c0556e67d7f8">

so for my use case - many languages but not many pages - it would greatly reduce build time if that dependency step could be reused after the first language is done. That would probably reduce my time by 90%+

**Comment by slorber** [comment]
@SubJunk you seem to have a lot of languages, wouldn't it be better to have one deployment per locale and parallelize a bit more instead of deploying sequentially? 

You can choose your deployment strategy according to your scalability needs.

> One thing I notice while watching the output logs, is that most of the time is spent re-compiling the dependencies over and over, for each language: <img alt="image" width="809" src="https://private-user-images.githubusercontent.com/584228/292699556-0bd08771-d074-4d70-8776-c0556e67d7f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3MTU4ODcsIm5iZiI6MTcwMzcxNTU4NywicGF0aCI6Ii81ODQyMjgvMjkyNjk5NTU2LTBiZDA4NzcxLWQwNzQtNGQ3MC04Nzc2LWMwNTU2ZTY3ZDdmOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyN1QyMjE5NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iN2UzZjRiYjFmOTI3ZmMzNjYzZTNmOTE2N2M4MWZiNDMwNDUzMGQ5ZWIyOTRiMWU2Y2JjZTBmZmY4NTg2ODhkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.q994cyori2WbD4UntbIR5oQSLREJl1K95-v6l_NoM8U">
> 
> so for my use case - many languages but not many pages - it would greatly reduce build time if that dependency step could be reused after the first language is done. That would probably reduce my time by 90%+

I plan to pre-transpile our official theme code so that we can skip it during site builds, but it probably requires creating some build tooling first to make it manageable.

Note that your mdx docs are also compiled with MDX, and this takes time, and here all locales will have a different input.

Webpack has a persistent caching system, and if you cache `node_modules/.cache` across CI runs it could speed up your build. 

However I wish a part of that cache was shared across multiple locales but last time I investigated webpack does not permit that. (https://github.com/webpack/webpack/issues/13034)

**Comment by mapachurro** [comment]
Hello all!

This is the Thread I've Been Looking For.

I'm in the process of trying to get our site live in production; unfortunately repo is still behind the company GitHub org, but: we've got seventeen locales, a total of ~1400 articles, on Docusaurus V3, and we've been hitting the GH Actions / Pages RAM issue hard.

I've tried a few of the fixes mentioned here and elsewhere, and it looks like the next step is webpack. 

Is there, at this point, an off-the-shelf process for switching to webpack, or do I need to engineer a custom solution, as above? 

Apologies if this is documented somewhere and I missed it.

**Comment by Ornias1993** [comment]
We over at TrueCharts have it set up for a few thousand of pages, using github actions.
(github.com/truecharts)

**Comment by societymartingale** [comment]
Switching to swc-loader reduced my Docusaurus build time by about 20%. This is for a site with 600 mdx files. Build time is 2.5 minutes on an Apple M3 pro and 4.5 minutes in a GitLab CI docker-based build.

**Comment by damageboy** [comment]
@societymartingale do you have an example for how this was done with docusarus?

**Comment by johnnyreilly** [comment]
https://johnnyreilly.com/faster-docusaurus-build-swc-loader

**Comment by damageboy** [comment]
@johnnyreilly I've just literally found it minutes before you posted, tried and unfortunately for me, it:

- failed for me right at the end:
```
   Error: Docusaurus server-side rendering could not render static page with path / because of error: React is not defined
    at render (main:915642:26)
    at async /Users/dmg/projects/docu3/node_modules/p-map/index.js:57:22 {
  [cause]: ReferenceError: React is not defined
      at 30128 (main:935773:36)
      at __webpack_require__ (main:1025163:42)
      at 7470 (main:935839:66)
      at __webpack_require__ (main:1025163:42)
      at main:901981:80
  }

Error: Unable to build website for locale en.
    at tryToBuildLocale (/Users/dmg/projects/docu3/node_modules/@docusaurus/core/lib/commands/build.js:55:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async mapAsyncSequential (/Users/dmg/projects/docu3/node_modules/@docusaurus/utils/lib/jsUtils.js:44:24)
    at async Command.build (/Users/dmg/projects/docu3/node_modules/@docusaurus/core/lib/commands/build.js:82:21) {
  [cause]: Error: Failed to compile due to Webpack errors.
  Error: Docusaurus server-side rendering could not render static page with path / because of error: React is not defined
      at /Users/dmg/projects/docu3/node_modules/@docusaurus/core/lib/webpack/utils.js:207:24
      at /Users/dmg/projects/docu3/node_modules/webpack/lib/MultiCompiler.js:596:14
      at processQueueWorker (/Users/dmg/projects/docu3/node_modules/webpack/lib/MultiCompiler.js:533:6)
      at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
}

[INFO] Docusaurus version: 3.1.1
Node version: v20.11.1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
yarn build  239.56s user 16.38s system 196% cpu 2:10.18 total
```

- (Assuming I find the issue and fix it) wasn't super fast compared to a normal build:
```
yarn build  265.45s user 18.76s system 201% cpu 2:21.16 total
```

Granted that 10% is not nothing, but also not 20%...

**Comment by Ornias1993** [comment]
YOu can also get a decent performance improvement by not processing `.md` files as `.mdx`

**Comment by damageboy** [comment]
> You can also get a decent performance improvement by not processing `.md` files as `.mdx`

Tips on how to get this done?

**Comment by slorber** [comment]
> YOu can also get a decent performance improvement by not processing `.md` files as `.mdx`

I'm also curious to know what you mean, maybe you are right but my intuition is that it does not have a significant impact since in both cases content is going to be compiled to React components.

---

Note, I'm actively working on perf optimizations for Docusaurus v3.2.

There are no breaking changes in the 3.x branch yet so if you can run a canary version of Docusaurus and let me know how it improves, I'd be curious to know how faster it is on your site.
https://docusaurus.io/community/canary

The last remaining bottleneck remains the Webpack compilation time, I'll look into that soon.

**Comment by societymartingale** [comment]
@damageboy, I added the following to package.json:
```
"@swc/core": "^1.4.6",
"swc-loader": "^0.2.6"
```

And added following to docusaurus.config.js:
```
webpack: {
    jsLoader: (isServer) => ({
      loader: require.resolve("swc-loader"),
      options: {
        jsc: {
          parser: {
            syntax: "typescript",
            tsx: true,
          },
          target: "es2019",
          transform: {
            react: {
              runtime: "automatic",
            },
          },
        },
        module: {
          type: isServer ? "commonjs" : "es6",
        },
      },
    }),
  },
```

As indicated above, I also disabled the last update author/time feature, as it wasn't scaling well. This saved another minute or two.

```
showLastUpdateAuthor: false,
showLastUpdateTime: false
```

**Comment by slorber** [comment]
> I also disabled the last update author/time feature, as it wasn't scaling well.

This major perf issue is fixed in canary (https://github.com/facebook/docusaurus/pull/9890) and will be released in v3.2

**Comment by andrewgbell** [comment]
3.2 for us has seen a major improvement in build times (thanks @slorber !). Running on Github actions, 4 core 16gb linux runner, we have the following

3.1 build time - 35 mins

3.2 build time - 23 mins
3.2 build time (esbuild) - 27 mins
3.2 build time (swc) - 21 mins

**Comment by Steveantor** [comment]
Why not ditch webpack for esbuild like vite? It's faster and better in every possible way.

**Comment by slorber** [comment]
@Steveantor it's easier said than done.

First of all, our plugin system has a [`configureWebpack`](https://docusaurus.io/docs/api/plugin-methods/lifecycle-apis#configureWebpack) hook, which means migrating away from Webpack would break most of our ecosystem that would need to update and provide a version compatibility matrix to their user. We also must ensure that an upgrade path is possible for plugin authors under the new bundler, at least for all major plugins.

I'm likely to adopt an incremental migration path thanks to [unplugin](https://github.com/unjs/unplugin), which is a more portable abstraction for bundler plugins that would probably be good enough for most Docusaurus plugins that only need a loader.

Also, [Vite is only using esbuild in some parts](https://vitejs.dev/guide/why#why-not-bundle-with-esbuild), and uses Rollup for bundling (v4 is based on SWC). And they are also working on Rolldown, a Rust port of Rollup. 

Due to how Docusaurus works, and our plugin system, it does not look like a good idea to use different bundling in dev/build modes.

Afaik esbuild supports live reloading but has no official for hot reloading. We don't want to refresh the browser when you edit JS or MD, this would make your page lose its state.

There's also Rspack which aims to be 100% retro-compatible with Webpack and according to this Bun benchmark, it's already quite faster than Webpack.

https://bun.sh/blog/bun-bundler

![image](https://github.com/facebook/docusaurus/assets/749374/07794862-e19f-4468-8983-e94190257acc)

Vercel is also actively working on [Turbopack](https://turbo.build/pack/docs/migrating-from-webpack), which also aims to be "mostly" compatible with Webpack, but less than Webpack.

So Rspack is for me the most suitable candidate in the short term, due to the constraints we have, and until other solutions become more mature. I'm likely to introduce ["future flags"](https://remix.run/blog/future-flags) in Docusaurus and let you swap Webpack with Rspack. It might not work for all third-party plugins (yet), but it should improve over time as Rspack fills the gap.


---

Note that bundling is a major bottleneck (mostly for "cold builds" with an empty Webpack cache, less for rebuilds) but is not the only performance issue we have in Docusaurus. I fixed some in v3.2 but I have ideas to improve other parts as well. 

Notably, I'm not sure the high memory consumption is related to the bundling phase, but rather the SSG phase.

**Comment by snake-py** [comment]
So I am also running into this issue, but my site is really small. I only have about three pages right now. It seems that the main issue is with CssMinimizerPlugin for me. 
![image](https://github.com/facebook/docusaurus/assets/59285387/e02347ae-38db-4320-86b8-9e1e18b6f9b6)

Is it possible to disable the minimizer?

**Comment by slorber** [comment]
You can try running with `USE_SIMPLE_CSS_MINIFIER=true docusaurus build` and see if it improves.

**Comment by Romej** [comment]
I would love to use docusaurus with rspack, we had many legacy CRA projects and we successfully migrated all of them to rspack without much efforts.

with rspack 1.0 around the corner, i am hoping this will be an option soon

**Comment by clainchoupi** [comment]
That may be a noob question, but how can I set the DOCUSAURUS_PERF_LOGGER variable ? 
I wanted to try it locally, so I tweaked a little my "npm run" commands but it does not work ^^

**Comment by OzakIOne** [comment]
> That may be a noob question, but how can I set the DOCUSAURUS_PERF_LOGGER variable ? I wanted to try it locally, so I tweaked a little my "npm run" commands but it does not work ^^

@clainchoupi I think it should be something like `DOCUSAURUS_PERF_LOGGER=true npm run start`

**Comment by clainchoupi** [comment]
I tried the following commands but none of them worked :P 

```
PS D:\WORK\TECH\docusaurus-interne> --DOCUSAURUS_PERF_LOGGER=true npm run start
==> DOCUSAURUS_PERF_LOGGER=true : Le terme «DOCUSAURUS_PERF_LOGGER=true» n'est pas reconnu comme nom d'applet de commande, fonction, fichier de      
script ou programme exécutable

PS D:\WORK\TECH\docusaurus-interne> DOCUSAURUS_PERF_LOGGER=true npm run start
==> Au caractère Ligne:1 : 3  --DOCUSAURUS_PERF_LOGGER=true npm run start   ~ Expression manquante après l'opérateur unaire « -- ».

PS D:\WORK\TECH\docusaurus-interne> npm run docusaurus build --DOCUSAURUS_PERF_LOGGER true
==> Build ok mais n'affiche pas les logs

PS D:\WORK\TECH\docusaurus-interne> npm run docusaurus build DOCUSAURUS_PERF_LOGGER true
==> [ERROR] [Error: ENOENT: no such file or directory, lstat 'D:\WORK\TECH\docusaurus-interne\DOCUSAURUS_PERF_LOGGER'] {

PS D:\WORK\TECH\docusaurus-interne> npm run docusaurus build DOCUSAURUS_PERF_LOGGER=true
==> [ERROR] [Error: ENOENT: no such file or directory, lstat 'D:\WORK\TECH\docusaurus-interne\DOCUSAURUS_PERF_LOGGER'] {
```

I'll try with other parameters to see :)

**Comment by slorber** [comment]
@clainchoupi each OS may have different ways to setup env variables (and it also depends on the shell you use on Windows afaik). But it's out of scope of Docusaurus to explain you how to set up env variables for your OS, you'll find plenty of tutorials online. The advise from @OzakIOne applies to Unix systems, not Windows (unless you use a compatible shell). 

If you don't know how to do that, you can try to use [cross-env](https://github.com/kentcdodds/cross-env), it should work in most contexts in an unified way.

I'm going to hide these discussions as off-topic because it's more about env variables than perf. Please open a dedicated discussion if needed, or ask support on Discord.

**Comment by slorber** [comment]
FYI Docusaurus v3.6 is entirely focused on performance - see the [Docusaurus Faster project](https://github.com/facebook/docusaurus/issues/10556)

The "faster" option is opt-in. It's a breaking change although most sites should be able to adopt it relatively easily.

Activating this option should have a significant impact on build time and memory consumption. 

You can try a Docusaurus canary right now and let me know how fast it is. It is relatively stable and ready to release, I'm just polishing the last details and writing docs before tagging v3.6.

**Comment by slorber** [comment]
Docusaurus v3.6 is out with the Docusaurus Faster options 🥳 🎉 

https://docusaurus.io/blog/releases/3.6

**Comment by Airkro** [comment]
`Docusaurus Faster` is still not stable.

Please consider using `options.experiments.lazyCompilation` of webpack. It makes everything faster when in dev mode.

Only need more `<Suspense>` to support it better.

**Comment by slorber** [comment]
@Airkro, it won't be stable until Docusaurus v4 because it's a breaking change anyway, Rspack is not 100% retrocompatible with webpack, and we can't guarantee a seamless upgrade for absolutely all of our sites. It doesn't mean it doesn't work well, and many large sites have already adopted it successfully. 

I'm also still working on performance improvements, so it may even be a bit too soon for Docusaurus v4 to use Docusaurus Faster by default, considering I'm still working on improvements and we need community feedback/adoption. If we turn on Docusaurus Faster by default, we'll likely provide an opt out to restore former behavior.


---

You can always tweak our default webpack client/server dev/prod configurations, thanks to the plugin lifecycle methods

```js
function myPlugin() {
  return {
    name: "my-plugin",
    configureWebpack() {
      return {
        experiments: {
          lazyCompilation: true,
        },
      };
    },
  };
}
```

Last time I tried using `lazyCompilation`, it didn't work well at all; the site couldn't build, and the errors were unhelpful in troubleshooting the problem. If you were able to run your site with this option, please let me know how or provide a repro. I'm looking to adopt it for Docusaurus.

**Comment by Airkro** [comment]
```mjs
if (process.env.WATCHING) { // custom detecting
  return {
    experiments: {
      lazyCompilation: {
        imports: true,
        entries: true
      }
    }
  }
}
```

Only using `lazyCompilation` in development mode (`docusaurus start`) can delay many unnecessary builds. It only works with `devServer`, not for building, but it still helps when the site becomes bigger and slower.

You may see page loading fail or blank pages(can be fixed by page refresh), but it will usually work fine.
If the docusaurus framework uses more `React.lazy + Suspense fallback`, it will have more obvious improvements.

**Comment by CMCDragonkai** [comment]
I came across this error https://github.com/facebook/docusaurus/issues/8329 in the Github CI/CD running `docusaurus build`.

What's actually scaling docusaurus's webpack phase memory usage? Is it the number of assets going through the webpack pipeline? If so increasing the heap memory will just fail once there's too many blog pages that feature some image or something. That's not really scalable.

What's the solution here, only use static URL paths and avoid the webpack pipeline? Or some way of making webpack discard un-used memory as much as possible?

**Comment by slorber** [comment]
@CMCDragonkai if you encounter a perf issue on a site it would be helpful to open a separate issue with all the details, otherwise it's really hard to do anything. If the site is https://polykey.com/docs then the first step would be to upgrade it to our latest release and turn on all the Docusaurus Faster options. I have no idea what you mean by "static URL paths".

**Comment by CMCDragonkai** [comment]
I shifted all static assets out of the webpack pipeline by referencing via pathname. I benchmarked webpacks memory usage too and it appears to be dropped by 1GiB. But it was getting bigger and bigger the more assets (images mainly I was referencing in docusaurus). So I actually think the webpack pipeline is only good for UI assets, for non-UI assets like content assets it has to use the static paths. And do direct cachebusting.

Yes I'm in docusaurus 3.9.
